src/loaders/umx_load.c

```diff
@@ -43,97 +42,374 @@ const struct format_loader libxmp_loader_umx = {
 
-		if (!memcmp(b, "Extended Module:", 16)) {
-			offset = i;
-			break;
-		}
-		if (id == MAGIC_IMPM) { 
-			offset = i;
-			break;
-		}
-		if (i > 44 && id == MAGIC_SCRM) { 
-			offset = i - 44;
-			break;
+	return t;	/* return type_name index */
+}
+
+static int read_typname(HIO_HANDLE *f, const struct upkg_hdr *hdr,
+			int idx, char *out)
+{
+	int i, s;
+	long l;
+	char buf[64];
+
+	if (idx >= hdr->name_count) return -1;
+	buf[63] = '\0';
+	for (i = 0, l = 0; i <= idx; i++) {
+		hio_seek(f, hdr->name_offset + l, SEEK_SET);
+		hio_read(buf, 1, 63, f);
+		if (hdr->file_version >= 64) {
+			s = *(signed char *)buf; /* numchars *including* terminator */
+			if (s <= 0 || s > 64) return -1;
+			l += s + 5;	/* 1 for buf[0], 4 for int32 name_flags */
+		} else {
+			l += (long)strlen(buf);
+			l +=  5;	/* 1 for terminator, 4 for int32 name_flags */
 		}
-		if (i > 1080 && id == MAGIC_M_K_) { 
-			offset = i - 1080;
+	}
+
+	strcpy(out, (hdr->file_version >= 64)? &buf[1] : buf);
+	return 0;
+}
+
+static int probe_umx   (HIO_HANDLE *f, const struct upkg_hdr *hdr,
+			int32 *ofs, int32 *objsize)
+{
+	int i, idx, t;
+	int32 s, pos;
+	long fsiz;
+	char buf[64];
+
+	idx = 0;
+	fsiz = hio_size(f);
+
+	/* Find the offset and size of the first IT, S3M or XM
+	 * by parsing the exports table. The umx files should
+	 * have only one export. Kran32.umx from Unreal has two,
+	 * but both pointing to the same music. */
+	if (hdr->export_offset >= fsiz) return -1;
+	memset(buf, 0, 64);
+	hio_seek(f, hdr->export_offset, SEEK_SET);
+	hio_read(buf, 1, 64, f);
+
+	get_fci(&buf[idx], &idx);	/* skip class_index */
+	get_fci(&buf[idx], &idx);	/* skip super_index */
+	if (hdr->file_version >= 60) idx += 4; /* skip int32 package_index */
+	get_fci(&buf[idx], &idx);	/* skip object_name */
+	idx += 4;			/* skip int32 object_flags */
+
+	s = get_fci(&buf[idx], &idx);	/* get serial_size */
+	if (s <= 0) return -1;
+	pos = get_fci(&buf[idx],&idx);	/* get serial_offset */
+	if (pos < 0 || pos > fsiz - 40) return -1;
+
+	if ((t = read_export(f, hdr, &pos, &s)) < 0) return -1;
+	if (s <= 0 || s > fsiz - pos) return -1;
+
+	if (read_typname(f, hdr, t, buf) < 0) return -1;
+	for (i = 0; mustype[i] != NULL; i++) {
+		if (!strcasecmp(buf, mustype[i])) {
+			t = i;
 			break;
 		}
 	}
-	
-	if (offset < 0)
-		return -1;
+	if (mustype[i] == NULL) return -1;
+	if ((t = get_objtype(f, pos, t)) < 0) return -1;
 
-	return 0;
+	*ofs = pos;
+	*objsize = s;
+	return t;
 }
 
-static int umx_load(struct module_data *m, HIO_HANDLE *f, const int start)
+static int32 probe_header (void *header)
 {
+	struct upkg_hdr *hdr;
+	unsigned char *p;
+	uint32 *swp;
 	int i;
-	uint8 buf[TEST_SIZE], *b = buf;
-	uint32 id;
 
-	LOAD_INIT();
+	/* byte swap the header - all members are 32 bit LE values */
+	p = (unsigned char *) header;
+	swp = (uint32 *) header;
+	for (i = 0; i < UPKG_HDR_SIZE/4; i++, p += 4) {
+		swp[i] = p[0] | (p[1] << 8) | (p[2] << 16) | (p[3] << 24);
+	}
```

Comment:
This seems questionable—`header` here is a stack allocated char array passed from `process_upkg` and I suspect it is not guaranteed to be DWORD aligned. Would it be less confusing (and safer) to just read these values directly into a `upkg_hdr` struct with `hio_read32l`?

Reply:
Done with the new commit: please review.